### ISSUE TITLE
override_host_set conflates different EDS health statuses

### ISSUE THREAD

**Issue by markdroth** [envoyproxy/envoy#23213]
This is something we discovered while in the process of implementing Envoy's stateful session affinity feature (see #16698) in gRPC.

From reading the code, it appears that there's a bug in Envoy's handling of the [`override_host_status` field](https://github.com/envoyproxy/envoy/blob/8e5c602e3cbafa3787269e94fe7fd0a81e62ca81/api/envoy/config/cluster/v3/cluster.proto#L672).  The field allows setting individual health statuses for which the host-override behavior should be honored.  However, internally, Envoy does not actually track all of the health statuses individually; in particular, it treats all of `DRAINING`, `UNHEALTHY`, and `TIMEOUT` as being exactly the same thing, and it treats `UNKNOWN` and `HEALTHY` as being the same thing (see [code](https://github.com/envoyproxy/envoy/blob/ea0eb2d5ea3f82250f3577c7ac5a898087d0799f/source/common/upstream/load_balancer_impl.cc#L529)).  So for example, if someone sets the `override_host_status` set to include `DRAINING` but not `UNHEALTHY` or `TIMEOUT`, Envoy will actually honor the host-override for all three of those states.  I think this is a bug; it seems very clear from the API what is actually intended here.

I think the underlying problem here is that Envoy does not internally record the actual EDS health status in the `Host` object; instead, it does a lossy conversion to `Host::Health`.  I suspect that the right fix is to have the `Host` object record the original EDS health status with full fidelity and then look at that when evaluating the `override_host_status` set.

CC @wbpcode @htuch @yanavlasov @sanjaypujare

**Comment by wbpcode** [comment]
> I suspect that the right fix is to have the Host object record the original EDS health status with full fidelity and then look at that when evaluating the override_host_status set.

Then how to handle the status set by active health checker or outlier detection? 🤔

**Comment by markdroth** [comment]
I would think that an active health checker failure or an outlier detection ejection would map to the EDS `UNHEALTHY` state.

**Comment by sanjaypujare** [comment]
or may be maintain 2 status fields as @wbpcode suggested?

**Comment by markdroth** [comment]
I think having two status fields would make the API more complicated, because then we would have to expose a way of controlling whether a given host will be used as an override based on the value of both fields.

The only status that is currently exposed in the xDS API is the EDS health status enum, and that enum is capable of representing the health states that Envoy drives from places other than EDS as well, so it seems to me that that's the right thing to use here.

**Comment by wbpcode** [comment]
/assign

Any way, we should solve this problem.

**Comment by htuch** [comment]
FTR I think it makes sense to keep track of the original EDS status in `Host`, so +1 to that. At the very least this would make sense when reporting endpoint status in endpoint dumps, but there are other advantages as Mark points out.

**Comment by markdroth** [comment]
I don't think this is fully fixed yet.

**Comment by wbpcode** [comment]
Yeah, an additional minor patch is necessary to match the override status with the new health status.

I am doing it now.

**Comment by wbpcode** [comment]
active
